### PR TITLE
docs: gateway hooks example project

### PR DESCRIPTION
**Pull request by pimeys**
Closes: GB-6898

### CODE REVIEW

File: examples/gateway-hooks/authorized-subgraph/src/schema.rs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+use async_graphql::{Context, Object, TypeDirective};
+
+#[TypeDirective(
+    name = "authorized",
+    location = "FieldDefinition",
+    location = "Object",
+    composable = "https://custom.spec.dev/extension/v1.0"
+)]
+fn authorized(rule: String, arguments: Option<String>) {}
```
Comment (Finistere) [inline examples/gateway-hooks/authorized-subgraph/src/schema.rs:11]: rule can be removed, could add `metadata`, but not sure of the type that should be used with async-graphql, with `serde_json::Value` it just serializes it to a string... maybe `async_graphql::Value`, but haven't tried it

Comment (Finistere) [reply examples/gateway-hooks/authorized-subgraph/src/schema.rs:11]: And you're never using it in the subgraph?

File: examples/gateway-hooks/authorized-subgraph/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    Router,
+};
+use schema::{QueryRoot, Users};
+use tokio::net::TcpListener;
+
+mod schema;
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    let schema = Schema::build(QueryRoot, EmptyMutation, EmptySubscription)
```
Comment (Finistere) [inline examples/gateway-hooks/authorized-subgraph/src/main.rs:15]: Should add `.enable_federation()` to expose `_service { sdl }`

File: examples/gateway-hooks/README.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+To run this example, you need the Grafbase Gateway version 0.4.0 or later, read more how to install it from:
+
+https://grafbase.com/docs/self-hosted-gateway
+
+Additionally, the following tools are needed:
+
+- Rust compiler ([install docs](https://www.rust-lang.org/learn/get-started))
+- Cargo component ([install docs](https://github.com/bytecodealliance/cargo-component?tab=readme-ov-file#installation))
+- A C compiler, such as clang (install based on your system, `cc` command is required)
+- A GraphQL client, such as
```
Comment (tomhoule) [inline examples/gateway-hooks/README.md:25]: Omission?

File: examples/gateway-hooks/README.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+For the advanced users using nix with flakes support, all the dependencies excluding the grafbase gateway can be installed with:
```
Comment (tomhoule) [inline examples/gateway-hooks/README.md:27]: Not in scope for this PR of course, but the gateway should be easy to package.

File: examples/gateway-hooks/demo-hooks/wasi_snapshot_preview1.reactor.wasm
Comment (tomhoule) [inline examples/gateway-hooks/demo-hooks/wasi_snapshot_preview1.reactor.wasm:1]: Is this intentionally committed?

**Review by tomhoule — APPROVED** [review]

File: examples/gateway-hooks/demo-hooks/wasi_snapshot_preview1.reactor.wasm
Comment (pimeys) [reply examples/gateway-hooks/demo-hooks/wasi_snapshot_preview1.reactor.wasm:1]: Yes, it is needed in the hooks until the next version of cargo-component is released. This fixes an issue if the user has a ton of env vars in scope, the reactor included in cargo-component would trap to OOM.

File: examples/gateway-hooks/README.md (hunk 4 above)
Comment (pimeys) [reply examples/gateway-hooks/README.md:27]: We could have it in the flake.nix of the repo and just include it in the devenv here.

File: examples/gateway-hooks/demo-hooks/wasi_snapshot_preview1.reactor.wasm
Comment (tomhoule) [reply examples/gateway-hooks/demo-hooks/wasi_snapshot_preview1.reactor.wasm:1]: :+1:

File: examples/gateway-hooks/README.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+Finally start the `grafbase-gateway`:
+
+```bash
+grafbase-gateway --schema federated-schema.graphql --config grafbase.toml
+```
+
+A successful start of the gateway will give the following output:
+
+```
+2024-07-02T17:33:17.242780Z  INFO Grafbase Gateway 0.3.2
```
Comment (fbjork) [inline examples/gateway-hooks/README.md:61]:
```suggestion
2024-07-02T17:33:17.242780Z  INFO Grafbase Gateway 0.4.0
```

File: examples/gateway-hooks/README.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+## Running the example
+
+First, start the subgraph in one terminal:
+
+```bash
+cd authorized-subgraph
+cargo run --release
+```
+
+Then, compile the WebAssembly hook functions into a wasm component in another terminal:
```
Comment (fbjork) [inline examples/gateway-hooks/README.md:42]:
```suggestion
Then, compile the WebAssembly hook functions into a Wasm component in another terminal:
```

**Pull request merged**